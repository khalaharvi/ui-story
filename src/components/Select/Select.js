import React from 'react';
import PropTypes from 'prop-types';
import _get from 'lodash.get';
import styled, { cx as emotionCx } from 'react-emotion';

import { createClassNames } from 'utils';
import { Caret as CaretIcon } from 'icons';

export const cx = createClassNames('custom-select');

const StyledSelect = styled('div')(({ disabled, theme }) => {
  let hoverStyles;
  let color;
  let borderStyles;
  const backgroundColor = _get(theme, 'buttons.secondary.backgroundColor');

  if (!disabled) {
    hoverStyles = {
      borderColor: _get(theme, 'buttons.secondary.hover.borderColor')
    };
  }

  color = _get(theme, 'buttons.secondary.color');

  borderStyles = {
    border: `1px solid ${_get(theme, 'buttons.secondary.borderColor')}`,
    borderColor: _get(theme, 'buttons.secondary.borderColor')
  };

  if (disabled) {
    color = _get(theme, 'buttons.secondary.disabled.color');

    borderStyles = {
      ...borderStyles,
      borderColor: _get(theme, 'buttons.secondary.disabled.borderColor')
    };
  }

  return {
    position: 'relative',

    backgroundColor,
    color,
    ...borderStyles,
    '&:hover': hoverStyles,
    height: 45,
    cursor: !disabled ? 'pointer' : 'not-allowed',

    svg: {
      position: 'absolute',
      display: 'block',
      top: '50%',
      right: 20,
      transform: 'translateY(-50%)',
      marginLeft: 5,
      width: 13,
      height: 6,
      stroke: 'currentColor'
    }
  };
});

const StyledHTMLSelect = styled('select')({
  cursor: 'pointer',
  width: '100%',
  height: '100%',
  margin: 0,
  padding: '0 38px 0 20px',
  background: 'none',
  outline: 0,
  appearance: 'none',
  border: 0,
  lineHeight: '43px',
  color: 'inherit',

  '&[disabled]': {
    cursor: 'not-allowed'
  }
});

export const Select = ({
  className,
  items,
  value,
  onChange
}) => (
  <StyledSelect className={emotionCx(cx(''), className)}>
    <StyledHTMLSelect
      className={cx('select')}
      value={value}
      onChange={e => onChange && onChange(e.target.value)}
    >
      {items.map(item => (
        <option
          className={cx('option')}
          key={item.value}
          disabled={item.disabled}
          value={item.value}
        >
          {item.label || item.value}
        </option>
      ))}
    </StyledHTMLSelect>
    <CaretIcon />
  </StyledSelect>
);

Select.propTypes = {
  className: PropTypes.string,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.string.isRequired
    })
  ).isRequired,
  value: PropTypes.string,
  onChange: PropTypes.func
};

export default Select;
