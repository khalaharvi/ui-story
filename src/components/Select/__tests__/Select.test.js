import React from 'react';
import { mount, shallow } from 'enzyme';

import { Select } from '../Select';

describe('<Select />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <Select
        onChange={() => null}
        items={[
          { value: 'index1', label: 'index name 1' },
          { value: 'index2', label: 'index name 2' },
          { value: 'index3', label: 'index name 3' },
          { value: 'index4', label: 'index name 4' }
        ]}
        value="index1"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with className', () => {
    const wrapper = shallow(
      <Select
        onChange={() => null}
        items={[
          { value: 'index1', label: 'index name 1' },
          { value: 'index2', label: 'index name 2' },
          { value: 'index3', label: 'index name 3' },
          { value: 'index4', label: 'index name 4' }
        ]}
        value="index1"
        className="custom-select-classname"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("calls 'onChange' prop on change", () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Select
        onChange={onChange}
        items={[
          { value: 'index1', label: 'index name 1' },
          { value: 'index2', label: 'index name 2' },
          { value: 'index3', label: 'index name 3' },
          { value: 'index4', label: 'index name 4' }
        ]}
        value="index1"
      />
    );

    const selectedValue = wrapper.find('select');
    expect(selectedValue.find('option')).toHaveLength(4);
    expect(
      selectedValue
        .find('option')
        .first()
        .text()
    ).toBe('index name 1');

    selectedValue
      .find('select')
      .simulate('change', { target: { value: 'index3' } });

    expect(onChange.mock.calls).toHaveLength(1);
    expect(onChange.mock.calls[0][0]).toEqual('index3');
  });

  it('should use value if no label provided', () => {
    const wrapper = mount(
      <Select
        onChange={() => null}
        items={[
          { value: 'index1' },
          { value: 'index2' },
          { value: 'index3' },
          { value: 'index4' }
        ]}
        value="index1"
      />
    );

    const selectedValue = wrapper.find('select');
    expect(selectedValue.find('option')).toHaveLength(4);
    expect(
      selectedValue
        .find('option')
        .first()
        .text()
    ).toBe('index1');
  });
});
