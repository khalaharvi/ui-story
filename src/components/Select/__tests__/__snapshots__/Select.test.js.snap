// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> renders correctly 1`] = `
<Styled(div)
  className="dh-custom-select"
>
  <Styled(select)
    className="dh-custom-select-select"
    onChange={[Function]}
    value="index1"
  >
    <option
      className="dh-custom-select-option"
      key="index1"
      value="index1"
    >
      index name 1
    </option>
    <option
      className="dh-custom-select-option"
      key="index2"
      value="index2"
    >
      index name 2
    </option>
    <option
      className="dh-custom-select-option"
      key="index3"
      value="index3"
    >
      index name 3
    </option>
    <option
      className="dh-custom-select-option"
      key="index4"
      value="index4"
    >
      index name 4
    </option>
  </Styled(select)>
  <Caret />
</Styled(div)>
`;

exports[`<Select /> renders correctly with className 1`] = `
<Styled(div)
  className="dh-custom-select custom-select-classname"
>
  <Styled(select)
    className="dh-custom-select-select"
    onChange={[Function]}
    value="index1"
  >
    <option
      className="dh-custom-select-option"
      key="index1"
      value="index1"
    >
      index name 1
    </option>
    <option
      className="dh-custom-select-option"
      key="index2"
      value="index2"
    >
      index name 2
    </option>
    <option
      className="dh-custom-select-option"
      key="index3"
      value="index3"
    >
      index name 3
    </option>
    <option
      className="dh-custom-select-option"
      key="index4"
      value="index4"
    >
      index name 4
    </option>
  </Styled(select)>
  <Caret />
</Styled(div)>
`;
